hasktorch/src/Torch/NN.hs

```diff
@@ -41,73 +43,114 @@ nextParameter = do
+data G
+instance Monad m => MonadState G m
+
+instance (HasForward modelA inA, HasForward modelB inB) => HasForwardProduct 'Stochastic 'Deterministic modelA inA modelB inB where
+  type BProduct 'Stochastic 'Deterministic modelA inA modelB inB = G -> ((B modelA inA, B modelB inB), G)
+  forwardProduct _ _ modelA inA modelB inB = \g -> let (outA, g') = forward modelA inA g in ((outA, forward modelB inB), g')
+
+instance (HasForward modelA inA, HasForward modelB inB) => HasForwardProduct 'Deterministic 'Stochastic modelA inA modelB inB where
+  type BProduct 'Deterministic 'Stochastic modelA inA modelB inB = G -> ((B modelA inA, B modelB inB), G)
+  forwardProduct _ _ modelA inA modelB inB = \g -> let (outB, g') = forward modelB inB g in ((forward modelA inA, outB), g')
+
+instance (HasForward modelA inA, HasForward modelB inB) => HasForwardSum 'Stochastic 'Deterministic modelA inA modelB inB where
+  type BSum 'Stochastic 'Deterministic modelA inA modelB inB = G -> (Maybe (Either (B modelA inA) (B modelB inB)), G)
+  forwardSum _ _ (Left modelA) (Left inA) = \g -> let (outA, g') = forward modelA inA g in (Just . Left $ outA, g')
+  forwardSum _ _ (Right modelB) (Right inB) = \g -> (Just . Right $ forward modelB inB, g)
+  forwardSum _ _ _ _ = \g -> (Nothing, g)
+
+instance (HasForward modelA inA, HasForward modelB inB) => HasForwardSum 'Deterministic 'Stochastic modelA inA modelB inB where
+  type BSum 'Deterministic 'Stochastic modelA inA modelB inB = G -> (Maybe (Either (B modelA inA) (B modelB inB)), G)
+  forwardSum _ _ (Left modelA) (Left inA) = \g -> (Just . Left $ forward modelA inA, g)
+  forwardSum _ _ (Right modelB) (Right inB) = \g -> let (outB, g') = forward modelB inB g in (Just . Right $ outB, g')
+  forwardSum _ _ _ _ = \g -> (Nothing, g)
+
+instance (HasForward modelA inA, HasForward modelB inB) => HasForwardSum' 'Stochastic 'Deterministic modelA inA modelB inB where
+  type BSum' 'Stochastic 'Deterministic modelA inA modelB inB = G -> (Either (B modelA inA) (B modelB inB), G)
+  forwardSum' _ _ (Left (modelA, inA)) = \g -> let (outA, g') = forward modelA inA g in (Left outA, g')
+  forwardSum' _ _ (Right (modelB, inB)) = \g -> (Right $ forward modelB inB, g)
+
+instance (HasForward modelA inA, HasForward modelB inB) => HasForwardSum' 'Deterministic 'Stochastic modelA inA modelB inB where
+  type BSum' 'Deterministic 'Stochastic modelA inA modelB inB = G -> (Either (B modelA inA) (B modelB inB), G)
+  forwardSum' _ _ (Left (modelA, inA)) = \g -> (Left $ forward modelA inA, g)
+  forwardSum' _ _ (Right (modelB, inB)) = \g -> let (outA, g') = forward modelB inB g in (Right outA, g')
+
+--
+-- Fully-stochastic instances
+--
+
+instance (HasForward modelA inA, HasForward modelB inB) => HasForwardProduct 'Stochastic 'Stochastic modelA inA modelB inB where
+  type BProduct 'Stochastic 'Stochastic modelA inA modelB inB = G -> ((B modelA inA, B modelB inB), G)
+  forwardProduct _ _ modelA inA modelB inB = do
```

Comment:
It's fine to use the state monad internally, but I don't think the types line up without calling runState before returning